Readme redo (#17)
casper-ecosystem/hello-world: README.md

# Hello World Session Code

This session code accepts a message string and stores it in the calling account under the `special_value` NamedKey.

> **Usage**: This session code expects a runtime argument named `message` of type string.

Reviewer: ```suggestion
**Usage**: This session code expects a runtime argument named `message` of the type `String`.
```

> **Tests**: There are two tests available to test the Hello World session code. The `should_store_hello_world` test verifies the happy path, where a string *hello world* is saved under the `special_value` NamedKey. The `should_error_on_missing_runtime_arg` test verifies that an error is displayed when the runtime argument is missing. 

Reviewer: ```suggestion
**Tests**: There are two tests available to test the Hello World session code. The `should_store_hello_world` test verifies the successful path, where a string *hello world* is saved under the `special_value` NamedKey. The `should_error_on_missing_runtime_arg` test verifies that an error is displayed when the runtime argument is missing. 
```

> The tests start by initializing the casper crates and creating a genesis account. Then the contract Wasm is loaded to the session_code object. The deploy_item object is created using the details like payment method, session code arguments, and account address. The deploy_item object is passed to the execute_request. Finally, the execution engine is invoked to process the execute_request. 

Reviewer: ```suggestion
The tests start by initializing the Casper crates and creating a genesis account. Then the contract Wasm is loaded to the session_code object. The deploy_item object is created using the details like payment method, session code arguments, and the account address. The deploy_item object is passed to the execute_request. Finally, the execution engine is invoked to process the execute_request. 
```

## Build and Test the Session Code 

### Set up the Rust toolchain
You need the Rust toolchain to develop smart contracts.
```bash
make prepare
```

> ### Compile smart contracts

Reviewer 2: ```suggestion
### Compile session code
```

> Compile WebAssembly (Wasm) files that will be used to deploy the session code.

Reviewer 2: ```suggestion
Compile the session code to WebAssembly (Wasm)
```

```bash
make build-contract
```

### Test
Run the tests suite.
```bash
make test
```

## Deploy the Hello World Session Code

You can deploy the Hello World session code on a local network using NCTL. For more information on how to run an NCTL network, see [Setting up an NCTL network](https://docs.casperlabs.io/dapp-dev-guide/building-dapps/setup-nctl/).

> This command provides a view into the faucet account details. The faucet is the default account created on the NCTL network.

Reviewer 2: ```suggestion
This command provides a view into the faucet account details. The faucet is the special pre-funded Account created on the NCTL network.
```

2022-06-21T10:11:49.196407 [INFO] [226848] NCTL :: faucet a/c purse         : uref-bc30bba7e0701b726e89c08154b6037a8bc4b0bb8635595fb254f6081f0da0b1-007
2022-06-21T10:11:49.198393 [INFO] [226848] NCTL :: faucet a/c purse balance : 999999999999999999999900000000000
2022-06-21T10:11:49.200185 [INFO] [226848] NCTL :: faucet on-chain account  : see below
{
  "api_version": "1.0.0",
  "block_header": null,
  "merkle_proof": "[2330 hex chars]",
  "stored_value": {
    "Account": {
      "account_hash": "account-hash-ba94ea5ab81adceb47ca4a0502f926633b5643d90beae3bdf3f55117ceaf2297",
      "action_thresholds": {
        "deployment": 1,
        "key_management": 1
      },
      "associated_keys": [
        {
          "account_hash": "account-hash-ba94ea5ab81adceb47ca4a0502f926633b5643d90beae3bdf3f55117ceaf2297",
          "weight": 1
        }
      ],
      "main_purse": "uref-bc30bba7e0701b726e89c08154b6037a8bc4b0bb8635595fb254f6081f0da0b1-007",
      "named_keys": [
        {
          "key": "uref-07ea6a04bc49f73d20b78f81d994cd324b60ea62046cc73e7a2030f4c25c2759-007",
          "name": "special_value"
        }
      ]
    }
  }
}
```

</details>

:::note

> Make a note of the NakedKey uref hash in the faucet account details.

Reviewer: ```suggestion
Make a note of the NamedKey URef hash in the faucet account details.
```

:::

### View the value stored in the NamedKey

> You can view the value stored in the NamedKey by using the uref hash assigned to the NamedKey. To do this, we will first find the state root hash of the network using the following command:

Reviewer: ```suggestion
You can view the value stored in the NamedKey by using the URef hash assigned to the NamedKey. To do this, we will first find the state root hash of the network using the following command:
```

```bash
casper-client get-state-root-hash --node-address http://localhost:11101
```

<details>
<summary>Sample output with state root hash</summary>

```json
{
"id": -7547762796950564402,
  "jsonrpc": "2.0",
  "result": {
    "api_version": "1.0.0",
    "state_root_hash": "6097c728c1af6179a491b5c3d143c49b032c9d6bb69be553729cb9f1489c3833"
  }
}
```

</details>

> Then use the uref hash and the state root hash to view the value stored in the NamedKey.

Reviewer: ```suggestion
Then use the URef hash and the state root hash to view the value stored in the NamedKey.
```
